storage: avoid having newly split replicas campaigning to init raft

spencerkimball opened the pull request:
Fixes #13202

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13208)
<!-- Reviewable:end -->

spencerkimball committed:
storage: avoid all replicas in newly split range campaigning on raft init

Fixes #13202

spencerkimball commented:
One point I don't understand here is that allocsim performance regresses if I allow the current leaseholder to initialize its Raft group, even when it doesn't campaign and doesn't unquiesce and send an empty proposal to wake the leader. I guess this is a question to you, @bdarnell: what side effects does creating the raw Raft node have?

petermattis commented:
> One point I don't understand here is that allocsim performance regresses if I allow the current leaseholder to initialize its Raft group, even when it doesn't campaign and doesn't unquiesce and send an empty proposal to wake the leader.

If the current lease holder creates its Raft group and then doesn't campaign, it has to wait for an election timeout before campaigning. If the lease holder doesn't create its Raft group then we'll go through the normal code paths. 

Btw, I was experimenting with this too and took a slightly different approach. Continue to create the Raft group, but only allow the current lease holder to campaign or allow anyone to campaign if we're within an election timeout of startup. 

```diff
diff --git a/pkg/storage/replica.go b/pkg/storage/replica.go
index a4dd6dd..d49e2e7 100644
--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -881,6 +881,10 @@ func (r *Replica) getLease() (*roachpb.Lease, *roachpb.Lease) {
 func (r *Replica) hasLease(ts hlc.Timestamp) bool {
        r.mu.Lock()
        defer r.mu.Unlock()
+       return r.hasLeaseLocked(ts)
+}
+
+func (r *Replica) hasLeaseLocked(ts hlc.Timestamp) bool {
        return r.mu.state.Lease.OwnedBy(r.store.StoreID()) &&
                r.leaseStatus(r.mu.state.Lease, ts, r.mu.minLeaseProposedTS).state == leaseValid
 }
@@ -1330,9 +1334,9 @@ func (r *Replica) assertStateLocked(reader engine.Reader) {
        }
 }

-// maybeInitializeRaftGroup check whether the internal Raft group has
-// not yet been initialized. If not, it is created and set to campaign
-// if possible.
+// maybeInitializeRaftGroup checks whether the internal Raft group has not yet
+// been initialized. If not and we don't hold the lease the Raft group is
+// created.
 func (r *Replica) maybeInitializeRaftGroup(ctx context.Context) {
        r.mu.Lock()
        defer r.mu.Unlock()
@@ -1342,7 +1346,12 @@ func (r *Replica) maybeInitializeRaftGroup(ctx context.Context) {
                r.raftMu.Lock()
                defer r.raftMu.Unlock()
                r.mu.Lock()
-               if err := r.withRaftGroupLocked(true /* shouldCampaign */, func(raftGroup *raft.RawNode) (bool, error) {
+               // Only campaign if we hold the lease or we're within an election timeout
+               // of node startup. After a split, only 1 of the 3 replicas will hold the
+               // lease allowing it to campaign.
+               shouldCampaign := r.hasLeaseLocked(r.store.Clock().Now()) ||
+                       !r.store.canCampaignIdleReplica()
+               if err := r.withRaftGroupLocked(shouldCampaign, func(raftGroup *raft.RawNode) (bool, error) {
                        return true, nil
                }); err != nil {
                        log.ErrEventf(ctx, "unable to initialize raft group: %s", err)
```

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/replica.go, line 1343 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbXfZx7t_sMu4TiVJZW:-KbXfZx7t_sMu4TiVJZX:b-r72m7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/29fb6725bd50c0eac25518ab4ded263a45d6c954/pkg/storage/replica.go#L1343)):*
> ```Go
> 	// comes up to date. If it's the lease owner, the command will
> 	// proceed (or the lease will be renewed).
> 	if r.mu.internalRaftGroup == nil && !r.mu.state.Lease.OwnedBy(r.store.StoreID()) {
> ```

Should we be checking that the lease is valid too (i.e. `Replica.hasLease`)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Your check for whether the store is within an election timeout of node startup won't work properly because `shouldCampaign` will be set to `false` in `withRaftGroupLocked` in the event that `internalRaftGroup == nil && r.store.canCampaignIdleReplica()`.

spencerkimball commented:
PTAL. I incorporated your idea as well.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replica.go, line 1343 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbXfZx7t_sMu4TiVJZW:-KbaLtEu6aXgEeNq_77y:b-3ccunb) ([raw file](https://github.com/cockroachdb/cockroach/blob/29fb6725bd50c0eac25518ab4ded263a45d6c954/pkg/storage/replica.go#L1343)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should we be checking that the lease is valid too (i.e. `Replica.hasLease`)?
</blockquote></details>

I think either way works fine, except my thinking was that ownership was more likely to designate a single replica than `Replica.hasLease`, which could have seen a lease or liveness expiration since the range was split.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-:-KbaPfiXnJdGtFGDFvxz:b3pvlm5)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ-2rKrKr2pIBHHIt:-KbaQ-2rKrKr2pIBHHIu:b677t1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1340)):*
> ```Go
> 	defer r.mu.Unlock()
> 	// If this replica hasn't initialized the Raft group and isn't the
> 	// lease owner, unquiesce and wake the leader to ensure replica
> ```

The `and isn't the lease owner` portion of this comment is out of date.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-KbaQ8Fw5RIEUeNcYPWP:b-gq96ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
> ```Go
> 		// Only campaign if this replica is the lease owner to avoid a
> 		// campaign storm after a recent split.
> 		shouldCampaign := r.mu.state.Lease.OwnedBy(r.store.StoreID())
> ```

We're doing this really early in processing. Is it possible for `r.mu.state.Lease` to be `nil`?

Do we need to find some way to incorporate the `r.store.canCampaignIdleReplica()` logic? If not, I think that logic and mechanism are no longer being used. The point of that logic was to allow fast campaigning of a lazily-loaded Raft group. Consider a cluster that has recently been restarted and an op arrives at an idle Range (i.e. one where the Raft group hasn't been initialized). In 2 out of 3 cases, the recipient will not be the previous lease holder and won't campaign and we'll have to wait an election timeout

Note that passing `shouldCampaign == true` in `defaultSubmitProposalLocked` has no effect now because we always hit `maybeInitializeRaftGroup` code path first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ-2rKrKr2pIBHHIt:-Kbadrm6Sjn3eKeaXFwu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1340)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `and isn't the lease owner` portion of this comment is out of date.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-Kbad3iYc63i5UYZwhAL:b5rf6w7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We're doing this really early in processing. Is it possible for `r.mu.state.Lease` to be `nil`?

Do we need to find some way to incorporate the `r.store.canCampaignIdleReplica()` logic? If not, I think that logic and mechanism are no longer being used. The point of that logic was to allow fast campaigning of a lazily-loaded Raft group. Consider a cluster that has recently been restarted and an op arrives at an idle Range (i.e. one where the Raft group hasn't been initialized). In 2 out of 3 cases, the recipient will not be the previous lease holder and won't campaign and we'll have to wait an election timeout

Note that passing `shouldCampaign == true` in `defaultSubmitProposalLocked` has no effect now because we always hit `maybeInitializeRaftGroup` code path first.
</blockquote></details>

`r.mu.state.Lease` cannot be nil, as it's always set when we load the replica state, which happens in `Replica.initLocked`.

The `r.store.canCampaignIdleReplica()` is being incorporated still. If this code path is encountered before the process has been alive for an election timeout, then we won't campaign, even if we specify `shouldCampaign=true` due to this replica being the owner of the lease.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-Kbag2jc9DRs2pa3PSAl:bnskudz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

`r.mu.state.Lease` cannot be nil, as it's always set when we load the replica state, which happens in `Replica.initLocked`.

The `r.store.canCampaignIdleReplica()` is being incorporated still. If this code path is encountered before the process has been alive for an election timeout, then we won't campaign, even if we specify `shouldCampaign=true` due to this replica being the owner of the lease.
</blockquote></details>

Right, but my point is what happens after the process has been alive longer than the election timeout. Then, only a request to the previous lease holder will cause a campaign. Before the addition of `maybeInitialRaftGroup`, any replica that received a request would campaign if it was part of an idle Raft group. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-KbahsHDMgZQ3z9F0zyN:bntwq7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Right, but my point is what happens after the process has been alive longer than the election timeout. Then, only a request to the previous lease holder will cause a campaign. Before the addition of `maybeInitialRaftGroup`, any replica that received a request would campaign if it was part of an idle Raft group. 
</blockquote></details>

Only if it either held the lease or if the lease was invalid or expired but owned by another replica so that this replica would proceed with lease acquisition. In that case, the lease acquisition would eventually end up with `shouldCampaign == true` in `defaultSubmitProposalLocked`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-KbanGuqN6pOXDajQ-Zf:b-r4zyxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Only if it either held the lease or if the lease was invalid or expired but owned by another replica so that this replica would proceed with lease acquisition. In that case, the lease acquisition would eventually end up with `shouldCampaign == true` in `defaultSubmitProposalLocked`.
</blockquote></details>

We only campaign in `withRaftGroupLocked` if `r.mu.internalRaftGroup == nil`. If a Replica is not the owner of the current lease, even if that lease is expired, we're not going to campaign here. And when we get to the call to `defaultSubmitProposalLocked` we'll have already initialized `r.mu.internalRaftGroup` and not campaign.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-Kbap0M4ad6i4mG4a_lG:bhprnkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We only campaign in `withRaftGroupLocked` if `r.mu.internalRaftGroup == nil`. If a Replica is not the owner of the current lease, even if that lease is expired, we're not going to campaign here. And when we get to the call to `defaultSubmitProposalLocked` we'll have already initialized `r.mu.internalRaftGroup` and not campaign.
</blockquote></details>

Ah I see now. We should really rename the `withRaftGroupLocked` parameter to `shouldCampaignOnCreation`.

Correct me if I'm wrong, but don't we immediately campaign for the RHS Raft group when we split a range?

Perhaps the correct thing to do is to campaign either if the lease is held by the current replica or else not held by any replica.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-Kbaq5BVJ67kZ7Mx5I84:b9tvsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
> Ah I see now. We should really rename the withRaftGroupLocked parameter to shouldCampaignOnCreation.

Sure. It made sense to me when I wrote it, but clearly that name is causing confusion.

> Correct me if I'm wrong, but don't we immediately campaign for the RHS Raft group when we split a range?

That's correct.

> Perhaps the correct thing to do is to campaign either if the lease is held by the current replica or else not held by any replica.

That sounds like it might work. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-:-Kbas8i71VRRKmgzfFsY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-Kbat4L9orgNp4u42xE9:b-1h3cgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> Ah I see now. We should really rename the withRaftGroupLocked parameter to shouldCampaignOnCreation.

Sure. It made sense to me when I wrote it, but clearly that name is causing confusion.

> Correct me if I'm wrong, but don't we immediately campaign for the RHS Raft group when we split a range?

That's correct.

> Perhaps the correct thing to do is to campaign either if the lease is held by the current replica or else not held by any replica.

That sounds like it might work. 
</blockquote></details>

OK, I made that suggested change and allocsim continues to have the correct behavior. We really must fashion a unittest which captures that behavior. Ideas?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-Kbav-UYWupTxDvGjtrp:b-dxx3d2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

OK, I made that suggested change and allocsim continues to have the correct behavior. We really must fashion a unittest which captures that behavior. Ideas?
</blockquote></details>

Perhaps we could have a `TestingKnobs` hook which fires when a replica tries to campaign. Then we can set up scenarios where the 3 replicas all receive a request simultaneously and verify the campaign behavior. We'd want to test with the various possibilities for the lease state (lease valid and held by one of the replicas, lease not valid but previously held by one of the replicas, lease not valid).

---

*[pkg/storage/replica.go, line 887 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaufVNlcDQiriki4GK:-KbaufVNlcDQiriki4GL:b-k7kv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b26e6d809fedad9152150290d62052cb47d687e/pkg/storage/replica.go#L887)):*
> ```Go
> }
> 
> func (r *Replica) hasLeaseLocked(ts hlc.Timestamp) bool {
> ```

You're not using `hasLeaseLocked` as far as I can tell.

---

*[pkg/storage/replica.go, line 1357 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbatwBb-HWDP5ipUzig:-KbatwBb-HWDP5ipUzih:bg0imeh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b26e6d809fedad9152150290d62052cb47d687e/pkg/storage/replica.go#L1357)):*
> ```Go
> 		// there's less chance of an election storm because this replica
> 		// will only campaign if it's been idle for >= election timeout,
> 		// so we can reasonably expect there's not much traffic to the range.
> ```

`s/not much/no/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-:-KbauuV1W9Cv2jwwv1LK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/replica.go, line 1351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaQ8Fw5RIEUeNcYPWO:-Kbb1CtapxUjfKzbKLoQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f99583535dd6fd157a0ed3c7b822bdddabdf7768/pkg/storage/replica.go#L1351)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Perhaps we could have a `TestingKnobs` hook which fires when a replica tries to campaign. Then we can set up scenarios where the 3 replicas all receive a request simultaneously and verify the campaign behavior. We'd want to test with the various possibilities for the lease state (lease valid and held by one of the replicas, lease not valid but previously held by one of the replicas, lease not valid).
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 887 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbaufVNlcDQiriki4GK:-Kbb1Aqe78Il3skwsFkE:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b26e6d809fedad9152150290d62052cb47d687e/pkg/storage/replica.go#L887)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're not using `hasLeaseLocked` as far as I can tell.
</blockquote></details>

Removed.

---

*[pkg/storage/replica.go, line 1357 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbatwBb-HWDP5ipUzig:-KbbLj07aytkPHNe2HK2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b26e6d809fedad9152150290d62052cb47d687e/pkg/storage/replica.go#L1357)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/not much/no/g`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_replica_test.go, line 1218 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbePfFRuitmqflDhuuJ:-KbePfFRuitmqflDhuuK:bmrhewx) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1218)):*
> ```Go
> 		syncutil.Mutex
> 		blockingCh chan struct{}
> 		campaigns  map[roachpb.ReplicaID]bool // map from ReplicaID -> whether the replica campaigned
> ```

nit: we use `map[T]struct{}` usually

---

*[pkg/storage/client_replica_test.go, line 1303 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbePVfV70N4caWI-ELL:-KbePVfV70N4caWI-ELM:b-hfpv1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1303)):*
> ```Go
> 			// Restart the cluster for lazy initialization of the raft group.
> 			mtc.restart()
> 			fmt.Println("\n\n********** Starting test case", i)
> ```

remove?

---

*[pkg/storage/client_replica_test.go, line 1316 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeQWopBJwZeO3w0toO:-KbeQWopBJwZeO3w0toP:b-mjt7fx) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1316)):*
> ```Go
> 			for _, s := range mtc.stores {
> 				errCh := make(chan error, 1)
> 				errChs = append(errChs, errCh)
> ```

why can't you use a single error channel?

---

*[pkg/storage/client_replica_test.go, line 1323 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeQAMEkffWQQfhSfrj:-KbeQAMEkffWQQfhSfrk:b-dkwnny) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1323)):*
> ```Go
> 					)
> 					errCh <- pErr.GoError()
> 				}(s, errCh)
> ```

nit: no need for `errCh` to be an argument

---

*[pkg/storage/client_replica_test.go, line 1331 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeQNJO8JJutLfydZ2m:-KbeQNJO8JJutLfydZ2n:b-6ztols) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1331)):*
> ```Go
> 				defer testState.Unlock()
> 				if c, ec := len(testState.campaigns), len(test.expCampaigns); c != ec {
> 					return errors.Errorf("have seen %d campaigns out of %d expected", c, ec)
> ```

won't this deadlock the test if it fails persistently?

---

*[pkg/storage/client_replica_test.go, line 1363 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeQiOjEchEiL_i7KqK:-KbeQiOjEchEiL_i7KqL:b-z1rzh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1363)):*
> ```Go
> 			// will not have a stray raft request init the raft group after
> 			// restart.
> 			time.Sleep(100 * time.Millisecond)
> ```

have you run this under stress?

---

*[pkg/storage/replica.go, line 473 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeOMjFXyJZm6QHcT6w:-KbeOMjFXyJZm6QHcT6x:b-hfpv1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/replica.go#L473)):*
> ```Go
> 
> 	if r.mu.internalRaftGroup == nil {
> 		log.Infof(ctx, "init of raft group")
> ```

remove?

---

*[pkg/storage/replica.go, line 1365 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeOhLE-L2oB7hJ5tQH:-KbeOhLE-L2oB7hJ5tQI:b-ksh2p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/replica.go#L1365)):*
> ```Go
> 			r.leaseStatus(r.mu.state.Lease, r.store.Clock().Now(), r.mu.minLeaseProposedTS).state != leaseValid
> 		st := r.leaseStatus(r.mu.state.Lease, r.store.Clock().Now(), r.mu.minLeaseProposedTS)
> 		log.Infof(ctx, "lease status: %d, lease: %s", st.state, st.lease)
> ```

did you mean to keep this? if so, move `st` up a line so it can be used in the definition of `shouldCampaignOnCreation`

---

*[pkg/storage/replica_range_lease.go, line 474 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeOBUWjFfaWvZ1z9zP:-KbeOBUWjFfaWvZ1z9zQ:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/replica_range_lease.go#L474)):*
> ```Go
> 		// Stop using the current lease.
> 		r.mu.minLeaseProposedTS = status.timestamp
> 		log.Infof(ctx, "****** RESETTING MIN LEASE: %s", status.timestamp)
> ```

ditto

---

*[pkg/storage/store.go, line 1835 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeO9ijCphLMpS7TjvG:-KbeO9ijCphLMpS7TjvH:b-vtqjze) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/store.go#L1835)):*
> ```Go
> 	// Copy the minLeaseProposedTS from the LHS.
> 	rightRng.mu.minLeaseProposedTS = r.mu.minLeaseProposedTS
> 	log.Infof(ctx, "****** SETTING MIN LEASE AFTER SPLIT: %s", r.mu.minLeaseProposedTS)
> ```

did you mean to remove this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Nice test!

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_replica_test.go, line 1218 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbePfFRuitmqflDhuuJ:-KbfSHYdIwF7ao_FxS5i:b-o3d7jg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1218)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: we use `map[T]struct{}` usually
</blockquote></details>

We also use `map[]bool`, though admittedly less frequently. The difference doesn't matter here.

---

*[pkg/storage/client_replica_test.go, line 1359 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbfRMVKpJtlr0SnJaj-:-KbfRMVKpJtlr0SnJaj0:b-pox3i3) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1359)):*
> ```Go
> 			testState.Unlock()
> 
> 			// HACK: sleep to process raft leader wakeup proposals. It is
> ```

Probably better to create a new `multiTestContext` per test-case. Doing so shouldn't be too expensive.

---

*[pkg/storage/replica.go, line 522 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbfRsLApnkOv3r3llxL:-KbfRsLApnkOv3r3llxM:bad4c3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/replica.go#L522)):*
> ```Go
> 			}
> 			// If testing knob is enabled, invoke OnCampaign function, taking care
> 			// to unlock the replica and re-lock in after invocation.
> ```

The unlock/re-lock changes the behavior of this method. Another goroutine can sneak in a make changes to the replica state. Do the tests require the replica to be unlocked?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-:-KbfTKjhBwjqW5rNNpfP:bds3i29)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_replica_test.go, line 1218 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbePfFRuitmqflDhuuJ:-KbjkITr5xEB4oc2lZ3k:b95obh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1218)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We also use `map[]bool`, though admittedly less frequently. The difference doesn't matter here.
</blockquote></details>

I prefer the use of bool here. It's considerably less verbose and the size of map values makes no difference for a test.

---

*[pkg/storage/client_replica_test.go, line 1303 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbePVfV70N4caWI-ELL:-KbjlEjLU_Y8lKr_v3p_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1303)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

remove?
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 1316 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeQWopBJwZeO3w0toO:-KbjlOVXbnxQeklUIkiD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1316)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why can't you use a single error channel?
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 1323 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeQAMEkffWQQfhSfrj:-KbjlP17Rhb14Ze5OtMb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1323)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: no need for `errCh` to be an argument
</blockquote></details>

Done.

---

*[pkg/storage/client_replica_test.go, line 1331 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeQNJO8JJutLfydZ2m:-KbjlRx7cGYwNuYbkPSB:b-apfqwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1331)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

won't this deadlock the test if it fails persistently?
</blockquote></details>

Moved the `close` statement out of the `testutils.SucceedsSoon` closure.

---

*[pkg/storage/client_replica_test.go, line 1359 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbfRMVKpJtlr0SnJaj-:-Kbjlr3QT8B9Rc-RfiHJ:b-x1lpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1359)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Probably better to create a new `multiTestContext` per test-case. Doing so shouldn't be too expensive.
</blockquote></details>

I tried that and it turns out that you still need the sleep, only you need it at the start of the test for the cases where we expect to have one or more replicas campaign.

---

*[pkg/storage/client_replica_test.go, line 1363 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeQiOjEchEiL_i7KqK:-Kbjm4q0dUU3LB4x1w5Q:bnn2t6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1363)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

have you run this under stress?
</blockquote></details>

For five minutes.

---

*[pkg/storage/replica.go, line 473 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeOMjFXyJZm6QHcT6w:-Kbjm9RgnbjiLBkNXLeZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/replica.go#L473)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

remove?
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 522 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbfRsLApnkOv3r3llxL:-KbjmDmIfnj3PRgnkIZr:b-u7gfx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/replica.go#L522)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The unlock/re-lock changes the behavior of this method. Another goroutine can sneak in a make changes to the replica state. Do the tests require the replica to be unlocked?
</blockquote></details>

Ugh, you're right, although it's unlikely to be a problem in tests. I'm going to get rid of this in favor of doing something more aggressive in helpers_test.go.

---

*[pkg/storage/replica.go, line 1365 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeOhLE-L2oB7hJ5tQH:-Kbjme3mJA2vY4b5Gudr:b-kj8xw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/replica.go#L1365)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

did you mean to keep this? if so, move `st` up a line so it can be used in the definition of `shouldCampaignOnCreation`
</blockquote></details>

No, removed.

---

*[pkg/storage/replica_range_lease.go, line 474 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeOBUWjFfaWvZ1z9zP:-KbjmossHAL3PwCuJCBV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/replica_range_lease.go#L474)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1835 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbeO9ijCphLMpS7TjvG:-KbjmsN5zWvnlQ0yiMx-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/store.go#L1835)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

did you mean to remove this?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_replica_test.go, line 1218 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KbePfFRuitmqflDhuuJ:-KbjzAycM3gkqpYzPWjW:b-ed97xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcc4214c9226a43456fe5d3a02f706332f070c87/pkg/storage/client_replica_test.go#L1218)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I prefer the use of bool here. It's considerably less verbose and the size of map values makes no difference for a test.
</blockquote></details>

Sure. My comment was motivated by semantic clarity, not concern for the size.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->

tamird commented:
please squash before merging

---

Reviewed 3 of 3 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_replica_test.go, line 1301 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KblYs0mNk1NZxIfPnPy:-KblYs0nf6UugnSvMzqj:b-l6mwxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/30819390ba03d7514c0b700fbae852e8457eb5ae/pkg/storage/client_replica_test.go#L1301)):*
> ```Go
> 	for i, test := range testCases {
> 		t.Run(test.desc, func(t *testing.T) {
> 			fmt.Println("\n\n********** Starting test case", i)
> ```

remove

---

*[pkg/storage/client_replica_test.go, line 1317 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KblZ5uSSKMcwaZ-yp5A:-KblZ5uSSKMcwaZ-yp5B:b-l6mwxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa0b2f6ed8214806781a58cdda2166a95588793/pkg/storage/client_replica_test.go#L1317)):*
> ```Go
> 
> 			// Restart the cluster for lazy initialization of the raft group.
> 			fmt.Println("\n\n********** Starting test case", i)
> ```

remove

---

*[pkg/storage/replica.go, line 517 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KblYuJeAZxY6ogg5BCv:-KblYuJeAZxY6ogg5BCw:b-l6mwxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/30819390ba03d7514c0b700fbae852e8457eb5ae/pkg/storage/replica.go#L517)):*
> ```Go
> 		if shouldCampaignOnCreation {
> 			log.VEventf(ctx, 3, "campaigning")
> 			log.Infof(ctx, "campaigning")
> ```

remove

---

*[pkg/storage/replica.go, line 525 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KblYvJfddb7BfXt17lt:-KblYvJfddb7BfXt17lu:b-eztgk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/30819390ba03d7514c0b700fbae852e8457eb5ae/pkg/storage/replica.go#L525)):*
> ```Go
> 			}
> 		} else {
> 			log.Infof(ctx, "initializing raft group but not campaigning")
> ```

remove or make less verbose

---

*[pkg/storage/replica.go, line 1362 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KblYxkZRZkGTtOgp_I_:-KblYxkZRZkGTtOgp_Ia:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/30819390ba03d7514c0b700fbae852e8457eb5ae/pkg/storage/replica.go#L1362)):*
> ```Go
> 		shouldCampaignOnCreation := r.mu.state.Lease.OwnedBy(r.store.StoreID()) ||
> 			r.leaseStatus(r.mu.state.Lease, r.store.Clock().Now(), r.mu.minLeaseProposedTS).state != leaseValid
> 		st := r.leaseStatus(r.mu.state.Lease, r.store.Clock().Now(), r.mu.minLeaseProposedTS)
> ```

ditto

---

*[pkg/storage/store.go, line 3115 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-KblYz4b0yS5DOLsijPU:-KblYz4b0yS5DOLsijPV:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/30819390ba03d7514c0b700fbae852e8457eb5ae/pkg/storage/store.go#L3115)):*
> ```Go
> 			r.setEstimatedCommitIndexLocked(req.Message.Commit)
> 		}
> 		log.Infof(ctx, "message: %+v", req.Message)
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-:-KblZDLq-lnrzCpHvUHu:bc2ilno)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Since I can only get failures running 5m on the CI infrastructure, I've been sending a lot of updates here. Disregard these changes please.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208#-:-Kbl_NyocltUax9ABSxz:be59duw)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13208)*
<!-- Sent from Reviewable.io -->
